File: selector.go
package main

import (
	"context"
	"fmt"
	"net"
	"os/exec"
	"regexp"
	"strconv"
	"sync"
	"time"
)

type ProbeSource string

const (
	ProbeSourceCache ProbeSource = "cache"
	ProbeSourceProbe ProbeSource = "probe"
)

type ProbeResult struct {
	Endpoint  string
	LatencyMS int64
	Source    ProbeSource
}

func selectBestEndpoint(ctx context.Context, host FinalHostConfig) (ProbeResult, error) {
	timeout := time.Duration(host.ProbeTimeoutMS) * time.Millisecond
	limit := host.ProbeConcurrency
	if limit <= 0 {
		limit = 1
	}

	semaphore := make(chan struct{}, limit)
	results := make(chan probeOutcome, len(host.Endpoints))
	var wg sync.WaitGroup

	for _, endpoint := range host.Endpoints {
		endpoint := endpoint
		wg.Add(1)

		go func() {
			defer wg.Done()

			select {
			case semaphore <- struct{}{}:
			case <-ctx.Done():
				results <- probeOutcome{Err: fmt.Errorf("%s:%d -> %w", endpoint, host.Port, ctx.Err())}
				return
			}
			defer func() { <-semaphore }()

			var latency time.Duration
			var err error
			switch host.SelectionMode {
			case SelectionModeLowestICMPLatency:
				latency, err = probeICMP(ctx, endpoint, timeout)
			default:
				latency, err = probeTCP(ctx, endpoint, host.Port, timeout)
			}
			if err != nil {
				results <- probeOutcome{Err: fmt.Errorf("%s:%d -> %w", endpoint, host.Port, err)}
				return
			}

			results <- probeOutcome{
				Result: ProbeResult{
					Endpoint:  endpoint,
					LatencyMS: latency.Milliseconds(),
					Source:    ProbeSourceProbe,
				},
			}
		}()
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	var best *ProbeResult
	var lastErr error
	for outcome := range results {
		if outcome.Err != nil {
			lastErr = outcome.Err
			continue
		}

		if best == nil || outcome.Result.LatencyMS < best.LatencyMS {
			result := outcome.Result
			best = &result
		}
	}

	if best == nil {
		if lastErr != nil {
			return ProbeResult{}, lastErr
		}
		return ProbeResult{}, fmt.Errorf("no reachable endpoint found")
	}

	return *best, nil
}

type probeOutcome struct {
	Result ProbeResult
	Err    error
}

func probeTCP(ctx context.Context, host string, port uint16, timeout time.Duration) (time.Duration, error) {
	address := fmt.Sprintf("%s:%d", host, port)
	dialer := &net.Dialer{Timeout: timeout}
	start := time.Now()

	conn, err := dialer.DialContext(ctx, "tcp", address)
	if err != nil {
		return 0, fmt.Errorf("connect failed: %w", err)
	}
	_ = conn.Close()

	return time.Since(start), nil
}

func probeICMP(ctx context.Context, host string, timeout time.Duration) (time.Duration, error) {
	timeoutSec := int(timeout / time.Second)
	if timeoutSec < 1 {
		timeoutSec = 1
	}

	commandCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	output, err := exec.CommandContext(commandCtx, "ping", "-c", "1", "-W", strconv.Itoa(timeoutSec), host).CombinedOutput()
	if err != nil {
		if commandCtx.Err() == context.DeadlineExceeded {
			return 0, fmt.Errorf("ping timeout")
		}
		return 0, fmt.Errorf("ping failed: %w", err)
	}

	latency, err := parsePingLatency(string(output))
	if err != nil {
		return 0, err
	}

	return latency, nil
}

var pingLatencyPattern = regexp.MustCompile(`time=([0-9]+(?:\.[0-9]+)?)\s*ms`)

func parsePingLatency(output string) (time.Duration, error) {
	matches := pingLatencyPattern.FindStringSubmatch(output)
	if len(matches) != 2 {
		return 0, fmt.Errorf("unable to parse ping latency")
	}

	value, err := strconv.ParseFloat(matches[1], 64)
	if err != nil {
		return 0, fmt.Errorf("unable to parse ping latency: %w", err)
	}

	return time.Duration(value * float64(time.Millisecond)), nil
}

File: src/sshe/selector.rs
use crate::sshe::config::{FinalHostConfig, SelectionMode};
use std::net::SocketAddr;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::net::{TcpStream, lookup_host};
use tokio::process::Command;
use tokio::sync::Semaphore;
use tokio::time;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProbeSource {
    Cache,
    Probe,
}

#[derive(Debug, Clone)]
pub struct ProbeResult {
    pub endpoint: String,
    pub latency_ms: u128,
    pub source: ProbeSource,
}

pub async fn select_best_endpoint(host: &FinalHostConfig) -> Result<ProbeResult, String> {
    let timeout = Duration::from_millis(host.probe_timeout_ms);
    let semaphore = Arc::new(Semaphore::new(host.probe_concurrency.max(1)));
    let mut tasks = Vec::with_capacity(host.endpoints.len());

    for endpoint in host.endpoints.iter().cloned() {
        let semaphore = Arc::clone(&semaphore);
        let selection_mode = host.selection_mode;
        let port = host.port;

        tasks.push(tokio::spawn(async move {
            let _permit = semaphore
                .acquire_owned()
                .await
                .map_err(|_| format!("{endpoint}:{port} -> semaphore closed"))?;

            let latency_ms = match selection_mode {
                SelectionMode::LowestTcpLatency => probe_tcp(&endpoint, port, timeout).await,
                SelectionMode::LowestIcmpLatency => probe_icmp(&endpoint, timeout).await,
            }
            .map_err(|err| format!("{endpoint}:{port} -> {err}"))?;

            Ok::<ProbeResult, String>(ProbeResult {
                endpoint,
                latency_ms,
                source: ProbeSource::Probe,
            })
        }));
    }

    let mut best: Option<ProbeResult> = None;
    let mut errors = Vec::new();

    for task in tasks {
        match task.await {
            Ok(Ok(result)) => match &best {
                Some(current) if current.latency_ms <= result.latency_ms => {}
                _ => best = Some(result),
            },
            Ok(Err(err)) => errors.push(err),
            Err(err) => errors.push(format!("probe task failed: {err}")),
        }
    }

    best.ok_or_else(|| {
        if errors.is_empty() {
            "no reachable endpoint found".to_string()
        } else {
            errors.join("; ")
        }
    })
}

async fn probe_tcp(host: &str, port: u16, timeout: Duration) -> Result<u128, String> {
    let addr = resolve_socket_addr(host, port).await?;
    let start = Instant::now();

    time::timeout(timeout, TcpStream::connect(addr))
        .await
        .map_err(|_| "connect timeout".to_string())?
        .map_err(|err| format!("connect failed: {err}"))?;

    Ok(start.elapsed().as_millis())
}

async fn resolve_socket_addr(host: &str, port: u16) -> Result<SocketAddr, String> {
    let addr_text = format!("{host}:{port}");
    let mut addrs = lookup_host(&addr_text)
        .await
        .map_err(|err| format!("resolve failed: {err}"))?;

    addrs
        .next()
        .ok_or_else(|| "no socket address resolved".to_string())
}

async fn probe_icmp(host: &str, timeout: Duration) -> Result<u128, String> {
    let timeout_sec = timeout.as_secs().max(1);
    let output = Command::new("ping")
        .args(["-c", "1", "-W", &timeout_sec.to_string(), host])
        .output()
        .await
        .map_err(|err| format!("failed to execute ping: {err}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let message = stderr.trim();
        if message.is_empty() {
            return Err("ping failed".to_string());
        }
        return Err(format!("ping failed: {message}"));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    parse_ping_latency(&stdout).ok_or_else(|| "unable to parse ping latency".to_string())
}

fn parse_ping_latency(stdout: &str) -> Option<u128> {
    let marker = "time=";
    let start = stdout.find(marker)? + marker.len();
    let tail = &stdout[start..];
    let end = tail.find(" ms").or_else(|| tail.find("ms"))?;
    let value = tail[..end].trim().parse::<f64>().ok()?;
    Some(value.round() as u128)
}

#[cfg(test)]
mod tests {
    use super::parse_ping_latency;

    #[test]
    fn parses_ping_output() {
        let sample = "64 bytes from 1.1.1.1: icmp_seq=1 ttl=57 time=12.7 ms\n\n--- 1.1.1.1 ping statistics ---";
        assert_eq!(parse_ping_latency(sample), Some(13));
    }
}
